common/src/worldgen.rs

```diff
@@ -62,13 +65,31 @@ impl NodeStateRoad {
     }
 }
 
+/// Contains a minimal amount of information about a node that can be deduced entirely from
+/// the NodeState of its parents.
+pub struct PartialNodeState {
+    /// This becomes a real horosphere only if it doesn't interfere with another higher-priority horosphere.
+    candidate_horosphere: Option<HorosphereNode>,
+}
+
+impl PartialNodeState {
+    pub fn new(graph: &Graph, node: NodeId) -> Self {
+        Self {
+            candidate_horosphere: HorosphereNode::new(graph, node),
+        }
+    }
+}
+
 /// Contains all information about a node used for world generation. Most world
-/// generation logic uses this information as a starting point.
+/// generation logic uses this information as a starting point. The `NodeState` is deduced
+/// from the `NodeState` of the node's parents, along with the `PartialNodeState` of the node
+/// itself and its "peer" nodes (See `peer_traverser`).
 pub struct NodeState {
     kind: NodeStateKind,
     surface: Plane,
     road_state: NodeStateRoad,
     enviro: EnviroFactors,
+    horosphere: Option<HorosphereNode>,
```

Comment:
I haven't exactly decided how we would want to support multiple different kinds of megastructures. We essentially need `NodeState` to be aware of every megastructure there is, but it seems silly to have a field for every type of megastructure when only one would be populated (barring future work of supporting multiple megastructures in the same node, like something inside a horosphere).

My current idea is to punt this design question for a future PR when we would actually have multiple megastructure types to choose from, as trying to figure this out now seems like a premature abstraction.

Reply:
Punting seems like a good call. I've always figured we'd have something like `Option<Box<dyn Megastructure>>` in the end, but getting the interface for that right will be difficult before we actually have multiple types. A good follow-up might be to implement a handful of hyperrogue classics to fill things out.